### PR TITLE
[REF] *: use hoot.click and hoot.drag in tours

### PR DESCRIPTION
**Pull request by pipu-odoo**
In this commit, click event and drag_and_drop_native events in web_tour/static/src/tour_service/tour_utils.js have been changed to use hoot events. Also, blur and focus utils have been removed from RunningTourActionHelper because there usages are not relevant and complicate the understanding of the API.

https://github.com/odoo/enterprise/pull/60855

### CODE REVIEW

File: addons/mass_mailing/static/tests/tours/mailing_editor.js [hunk 1]
```diff
@@ -18,14 +18,24 @@ registry.category("web_tour.tours").add('mailing_editor', {
 }, {
     content: 'choose the theme "empty" to edit the mailing with snippets',
     trigger: '[name="body_arch"] :iframe #empty',
+    run() {
+        //You should use hoot.click() but without target.focus(); in triggerFocus(); in events.js
```
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/mass_mailing/static/tests/tours/mailing_editor.js:22]: OMG ... forgot to remove this comment...

File: addons/website/static/tests/tours/website_text_font_size.js [hunk 2]
```diff
@@ -88,6 +88,10 @@ function getFontSizeTestSteps(fontSizeClass) {
         }, {
             content: `Remove the text snippet containing the text with class ${fontSizeClass}`,
             trigger: `.oe_snippet_remove`,
+            async run(helpers) {
+                helpers.click();
+                await new Promise((r) => setTimeout(r, 300));
+            },
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/website_text_font_size.js:94]: Why do we need this? It's deserving a comment, but maybe it's going to be removed following the other review discussion

File: addons/website/static/tests/tours/snippet_empty_parent_autoremove.js [hunk 3]
```diff
@@ -6,6 +6,9 @@ function removeSelectedBlock() {
     return {
         content: "Remove selected block",
         trigger: '#oe_snippets we-customizeblock-options:nth-last-child(3) .oe_snippet_remove',
+        run() {
+            this.anchor.click();
+        }
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/snippet_empty_parent_autoremove.js:11]: Can't this just be removed? Isn't click default behavior? It has to be explicitly marked as click now, isn't it? But is this way the shortest ? Seems verbose for just a click
`run: "click"` ?

File: addons/website/static/src/js/tours/tour_utils.js [hunk 4]
```diff
@@ -286,7 +287,15 @@ function dragNDrop(snippet, position = "bottom") {
...
-        run: "drag_and_drop_native :iframe #wrapwrap > footer",
+        async run(helpers) {
+            const target = snippet.target || `:iframe div[data-editor-message="DRAG BUILDING BLOCKS HERE"]:last`;
+            await helpers.drag_and_drop(target, {
+                position: {
+                    bottom: 0,
+                },
+                relative: true,
+            });
+        },
```
Comment (rdeodoo) [inline addons/website/static/src/js/tours/tour_utils.js:298]: Why this change

File: addons/website/static/tests/tours/link_tools.js [hunk 5]
```diff
@@ -106,7 +106,7 @@ wTourUtils.registerWebsitePreviewTour('link_tools', {
     {
         content: "Set URL.",
         trigger: '.o_we_customize_panel we-input:contains("Your URL") input',
-        run: "edit odoo.com && blur (we-title:contains(Your URL))",
+        run: "edit odoo.com && click(we-title:contains(Your URL))",
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/link_tools.js:109]: `&& click` to remove? next step is clicking on another img, it should be enough

File: addons/website/static/tests/tours/link_tools.js [hunk 6]
```diff
@@ -176,6 +191,16 @@ wTourUtils.registerWebsitePreviewTour('link_tools', {
         isCheck: true,
     },
     // 7. Create new a link from a URL-like text.
+    {
+        content: "click on Mega menu to hidden Mega menu",
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/link_tools.js:195]:
```suggestion
        content: "click on Mega menu item to hide Mega menu content",
```
I guess that's what you meant?

File: addons/website/static/tests/tours/link_tools.js [hunk 7]
```diff
@@ -240,6 +266,7 @@ wTourUtils.registerWebsitePreviewTour('link_tools', {
     {
         content: "Change URL into an email address",
         trigger: "#o_link_dialog_url_input",
+        extra_trigger: "div#oe_snippets:not(div.o_we_ui_loading)",
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/link_tools.js:269]: That seems most likely like a workaround to a possible race condition in our code @qsm-odoo , weird to have to add this everywhere in this tour.
From what I understood so far, the loading is about the autocomplete URL loading:
image

Seems very fragile to rely on that

File: addons/website/static/tests/tours/multi_edition.js [hunk 8]
```diff
@@ -21,6 +21,7 @@ wTourUtils.registerWebsitePreviewTour('website_multi_edition', {
     wTourUtils.dragNDrop({
         id: 's_hr',
         name: 'Separator',
+        target: `:iframe #wrapwrap footer`,
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/multi_edition.js:24]: Isn't this supposed to be the default behavior of dragNDrop?

File: addons/website/static/tests/tours/link_tools.js [hunk 9]
```diff
@@ -371,7 +408,8 @@ wTourUtils.registerWebsitePreviewTour('link_tools', {
     {
         content: "Verify that the link label input does not contain ZWS",
         trigger: "#o_link_dialog_label_input:value('Contact Us')",
+        extra_trigger: "div#oe_snippets:not(div.o_we_ui_loading)",
         isCheck: true,
     },
-    ...wTourUtils.clickOnSave(),
+    ...wTourUtils.clickOnSave("bottom", 20000),
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/link_tools.js:414]: Is this tour particularly slow? We generally tend to never increase timeout without a very good explanation of why, I am surprise to see it lost in the diff

File: addons/website/static/tests/tours/snippet_popup_and_scrollbar.js [hunk 10]
```diff
@@ -67,48 +68,62 @@ wTourUtils.registerWebsitePreviewTour("snippet_popup_and_scrollbar", {
...
+        content: "Open the Popup that has backdrop.",
+        trigger: ".o_we_invisible_el_panel .o_we_invisible_entry:first",
     },
     checkScrollbar(false),
     wTourUtils.goBackToBlocks(),
+    {
+        content: "Check that 'Blocks' tab is active",
+        trigger: ".o_snippet_search_filter:not(.d-none)",
+        isCheck: true,
+    },
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/snippet_popup_and_scrollbar.js:94]: This shouldn't be needed, as the trigger (snippet `s_media_list`) in next step is not supposed to exist / be found until this is done?

File: addons/website/static/tests/tours/snippet_popup_and_scrollbar.js [hunk 11]
```diff
@@ -67,48 +68,62 @@ wTourUtils.registerWebsitePreviewTour("snippet_popup_and_scrollbar", {
...
         trigger: ".o_we_customize_panel we-customizeblock-options:contains('Popup') we-button.oe_snippet_remove:first",
-        in_modal: false,
         run: "click",
     },
     checkScrollbar(true),
+    {
+        content: "Click on the s_cookie snippet",
+        trigger: ":iframe .s_popup .modal section:first",
+        run: "click",
+    },
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/snippet_popup_and_scrollbar.js:116]: Why do we need this? What is preventing to click on the "Blocks" tab?

File: addons/website/static/tests/tours/snippet_popup_and_scrollbar.js [hunk 12]
```diff
@@ -67,48 +68,62 @@ wTourUtils.registerWebsitePreviewTour("snippet_popup_and_scrollbar", {
...
     wTourUtils.goBackToBlocks(),
+    {
+        content: "Check that 'Blocks' tab is active",
+        trigger: ".o_snippet_search_filter:not(.d-none)",
+        isCheck: true,
+    },
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/snippet_popup_and_scrollbar.js:122]: Same here, seems useless, next step can't be found until the tab is active?

File: addons/website/static/tests/tours/snippet_popup_and_scrollbar.js [hunk 13]
```diff
@@ -117,8 +132,9 @@ wTourUtils.registerWebsitePreviewTour("snippet_popup_and_scrollbar", {
     {
         content: "Duplicate the Media List snippet",
         trigger: ".o_we_customize_panel we-customizeblock-options:contains('Media List') we-button.oe_snippet_clone:first",
-        in_modal: false,
-        run: "click",
+        run() {
+            this.anchor.click();
+        }
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/snippet_popup_and_scrollbar.js:137]: I am probably missing something obvious but why can't it be `run: "click",`?

File: addons/website/static/tests/tours/website_form_editor.js [hunk 14]
```diff
@@ -60,6 +60,9 @@
         return [{
             content: "Open the select",
             trigger: `we-select:has(we-button[${data}]) we-toggler`,
+            run() {
+                this.anchor.click();
+            }
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/website_form_editor.js:65]: same here, can't it be `run: "click",`?

File: addons/website/static/tests/tours/website_form_editor.js [hunk 15]
```diff
@@ -785,15 +788,29 @@
...
+            run(helpers) {
+                this.anchor.click();
+            }
         }, {
             content: "Choose first checkbox as condition item",
             trigger: 'we-button[data-set-visibility-dependency="Checkbox 1"]',
+            async run(helpers) {
+                helpers.click();
+                await new Promise((r) => setTimeout(r, 300));
+            }
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/website_form_editor.js:800]: ?

File: addons/website/static/tests/tours/website_form_editor.js [hunk 16]
```diff
@@ -785,15 +788,29 @@
...
+            run(helpers) {
+                this.anchor.click();
+            }
         }, {
             content: "Choose 'not equal to' comparator",
             trigger: 'we-button[data-select-data-attribute="!selected"]',
+            async run(helpers) {
+                helpers.click();
+                await new Promise((r) => setTimeout(r, 300));
+            }
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/website_form_editor.js:813]: ?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Quick one flying over it, I saw that the enterprise PR got r+ already so I did not dig a lot before commenting

File: addons/website/static/tests/tours/link_tools.js (hunk 5 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/link_tools.js:109]: I just modify the blur typo by click, I haven't added any action

File: addons/website/static/tests/tours/link_tools.js (hunk 9 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/link_tools.js:414]: The end of this tour is particularly slow yes. I've try with 10000 and didn't work on localhost.

File: addons/website/static/tests/tours/snippet_popup_and_scrollbar.js (hunk 13 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/snippet_popup_and_scrollbar.js:137]: Otherwise the tour fails. It could be perfect to use it. To be investigated...

File: addons/website/static/tests/tours/website_form_editor.js (hunk 14 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/website_form_editor.js:65]: Otherwise the tour fails. It could be perfect to use it. To be investigated...

File: addons/website/static/tests/tours/website_text_font_size.js [hunk 17]
```diff
@@ -61,7 +61,7 @@ function getFontSizeTestSteps(fontSizeClass) {
         }, {
             content: `Change the setting value of ${fontSizeClass}`,
             trigger: `[data-variable="${classNameInfo.get(fontSizeClass).scssVariableName}"] input`,
-            run: `edit ${classNameInfo.get(fontSizeClass).end} && blur`,
```
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/website/static/tests/tours/website_text_font_size.js:64]: As discussed, this is not the scope of this PR. I can add a task in our pipe to check each edit && click (outside).

File: addons/website/static/tests/tours/website_text_font_size.js (hunk 2 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/website_text_font_size.js:94]: I'll add a todo in a seperate commit

File: addons/website/static/tests/tours/snippet_empty_parent_autoremove.js (hunk 3 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/snippet_empty_parent_autoremove.js:11]: Done

File: addons/website/static/tests/tours/link_tools.js [hunk 18]
```diff
@@ -225,8 +242,9 @@ wTourUtils.registerWebsitePreviewTour('link_tools', {
     {
         content: "Change it back http",
         trigger: "#o_link_dialog_url_input",
+        extra_trigger: "div#oe_snippets:not(div.o_we_ui_loading)",
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/link_tools.js:245]: Can you add a `TODO` on all those `o_we_ui_loading` extra loading? Those should not be here, it's surely due to a real race condition in our code
As discussed here https://github.com/odoo/odoo/pull/162031/commits/a4ed03b97e96a3b228b5a48d9c5ab4dc138950b7#r1572365856

File: addons/website/static/tests/tours/link_tools.js [hunk 19]
```diff
@@ -328,9 +351,16 @@ wTourUtils.registerWebsitePreviewTour('link_tools', {
     },
     */
     // 11. Pick a URL with auto-complete
+    {
+        content: "Wait the sidebar is openend",
+        trigger: `we-title:contains(Inline text)`,
+        extra_trigger: `input#o_link_dialog_url_input`,
+        isCheck: true,
+    },
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/link_tools.js:359]: Also add a TODO here, this is imo a bug in the hoot framework, I don't see why we need a step to wait for `input#o_link_dialog_url_input` as it's already the trigger of next step

**Review by rdeodoo — APPROVED** [review]
I am already approving here the website part, but could you just:
- add the missing TODO 
- add a space before the `TODO`
- create a task in RD Website referring to this PR, explaining that all TODO in second commit have to be checked after the merge

@qsm-odoo Feel free to jump in in case you don't agree here, but @pipu-odoo would like to have this merged ASAP and look at the website TODO afterward.
Maybe someone from the team could help him figuring why are those new steps or trigger needed.

File: addons/website/static/tests/tours/snippet_images_wall.js [hunk 20]
```diff
@@ -58,7 +58,7 @@ selectSignImageStep,
 }, {
     content: "Change the link of the image",
     trigger: ".snippet-option-ReplaceMedia [data-set-url] input",
-    run: "edit /contactus && blur",
+    run: "edit /contactus && click body",
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/snippet_images_wall.js:61]: Also add a TODO here, before your change a blur or click body was not needed, see here before the HOOT pr:
https://github.com/odoo/odoo/blame/3a4883bde4f28b25d210fc339b0f98ad603fdb82/addons/website/static/tests/tours/snippet_images_wall.js#L61

Could you please check all `&& click` added on website files and add a `TODO` on the ones where you added it yourself, which was not there initially (before https://github.com/odoo/odoo/commit/29bd1ddbd4fd26afbbe709125200be20a0262967)?

The one where it was `text_blur`, it's ok to keep uncommented.

**Review by rdeodoo — COMMENTED** [review]
@qsm-odoo Following my previous message, also note that about all the `& click`, `& blur` here, there is two types:
1. The ones which are the adaptation of `text_blur`, I guess we could triple check those at some point, but it's unrelated to HOOT PRs
2. The ones which were introduced, as the one I comment here, which should be removed or commented (to be removed later).

Also worth noting, `change` and `blur` events were apparently not properly triggered in HOOT tour framework when a step is inside an input and that the step after is clicking on another element.

It should be the case now, as @detrouxdev explained that to @pipu-odoo which made some adaptation so the hoot framework reflects that correctly.

File: addons/website/static/tests/tours/link_tools.js (hunk 19 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/link_tools.js:359]: No longer needed

File: addons/website/static/tests/tours/snippet_images_wall.js (hunk 20 above)
Comment (qsm-odoo) [reply addons/website/static/tests/tours/snippet_images_wall.js:61]: Thanks @rdeodoo for checking all of this but I don't think we should go ahead with adding a bunch of TODO @pipu-odoo ... unless you think you will handle them quickly after the merge? :shrug: Or do we have any lead as to why adding a bunch of "click body" are necessary with the new system? Otherwise this may be a big hint that something is not perfect with the system in the first place no? :grimacing:

> Also worth noting, change and blur events were apparently not properly triggered in HOOT tour framework when a step is inside an input and that the step after is clicking on another element.
> 
> It should be the case now, as @detrouxdev explained that to @pipu-odoo which made some adaptation so the hoot framework reflects that correctly.

I am curious: how could it **not** be the case? Could you explain the changes that were made in this new system maybe? Apart from the syntax, why is the behavior different?

Thanks :pray:

File: addons/website/static/tests/tours/link_tools.js (hunk 18 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/link_tools.js]: I was able to remove a majority of them

File: addons/website/static/tests/tours/snippet_images_wall.js (hunk 20 above)
Comment (pipu-odoo) [reply addons/website/static/tests/tours/snippet_images_wall.js]: Hello, I think that a little more precision should actually be given for the edit() helper.
Unlike previously implemented in runningActionTourHelper's edit helper, hoot.edit() "only" clears the input and fills it (keydown, keyup). There are no change or input events that are triggered in hoot.edit(). To trigger a blur or change event, you must either press Enter, or click outside, or press Tab, etc.
In most of case, the next "click" action in a step tour is enough.
in a previous PR, I therefore replaced all text_blur with edit && click. In some places, edit alone was not enough. However, hoot.edit() has been improved and fixed in the meantime.

Comment (pipu-odoo) [reply addons/website/static/tests/tours/snippet_images_wall.js]: indeed. https://github.com/odoo-dev/odoo/blob/4169c4d698328ccb494b53ec1a1acf33d88a79ba/addons/web/static/lib/hoot-dom/helpers/events.js#L1068
